app/services/user_locale_setter.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+class UserLocaleSetter
+  def initialize(current_user, params_locale, cookies)
+    @current_user = current_user
+    @params_locale = params_locale
+    @cookies = cookies
+  end
+
+  def set_locale
+    save_locale_from_params
+
+    I18n.locale = valid_current_locale
+  end
+
+  def ensure_valid_locale_persisted
+    return unless current_user && !available_locale?(current_user.locale)
+
+    current_user.update!(locale: valid_current_locale)
+  end
+
+  def self.valid_locale_for_user(user)
```

Comment:
maybe valid_locale_for(user) ?

Reply:
there's repetition between this method and valid_current_locale.
The problem is that this is called from the mailed that can be used in delayedjob where there are no cookies.
I wonder if we could just add a if cookies on line 56 below and just reuse the methods.
That would mean remove this method and just make valid_current_locale public.

Reply:
Yes, I pulled it into the service but it's used in contexts where calling cookies will throw an error. I'll tidy this up a bit more :+1: